File: .leetcode/213.打家劫舍-ii.cpp
/*
 * @lc app=leetcode.cn id=213 lang=cpp
 *
 * [213] 打家劫舍 II
 */
#include<bits/stdc++.h>
using namespace std;

// @lc code=start
class Solution {
public:
    int rob(vector<int>& nums) {
        if(nums.size()==1)return nums[0];
        else{
            vector<int>nums1(nums.begin(),nums.end()-1);
            vector<int>nums2(nums.begin()+1,nums.end());
            return max(robCal(nums1),robCal(nums2));
        }
    }
    int robCal(vector<int>& nums) {
        if(nums.size()==1)return nums[0];
        if(nums.size()==2)return max(nums[0],nums[1]);
        if(nums.size()==3)return max(nums[0]+nums[2],nums[1]);
        vector<int>dp(nums.size(),0);
        dp[0] = nums[0];
        dp[1] = max(nums[0],nums[1]);
        dp[2] = max(nums[0]+nums[2],nums[1]);
        for(int i = 3;i<nums.size();i++){
            dp[i]=max(dp[i-3]+nums[i],dp[i-2]+nums[i]);
        }
        for(int i = 0;i<nums.size();i++){
            cout<<dp[i]<<" ";
        }
        cout<<endl;
        return max(dp[nums.size()-1],dp[nums.size()-2]);
    }
};
// @lc code=end

File: .leetcode/213.打家劫舍-ii.java
/*
 * @lc app=leetcode.cn id=213 lang=java
 *
 * [213] 打家劫舍 II
 */

// @lc code=start
class Solution {
    public int rob(int[] nums) {
        int n = nums.length;
        if(n == 0)return 0;
        if(n == 1)return nums[0];
        if(n == 2)return Math.max(nums[0], nums[1]);
        int[] dp1 = new int[n - 1];
        int[] dp2 = new int[n - 1];
        for(int i = 0;i < n - 1;i++){
            dp1[i] = Integer.MIN_VALUE;
            dp2[i] = Integer.MIN_VALUE;
        }
        dp1[0] = nums[0];
        dp1[1] = nums[1];
        dp2[0] = nums[1];
        dp2[1] = nums[2];
        for(int i = 0;i < n - 1;i++){
            if(i + 2 < n - 1)dp1[i + 2] = Math.max(dp1[i + 2], dp1[i] + nums[i + 2]);
            if(i + 3 < n - 1){
                dp1[i + 3] = Math.max(dp1[i + 3], dp1[i] + nums[i + 3]);
            }
        }
        for(int i = 1;i < n;i++){
            if(i + 2 < n){
                dp2[i + 1] = Math.max(dp2[i + 1], dp2[i - 1] + nums[i + 2]);
            }
            if(i + 3 < n){
                dp2[i + 2] = Math.max(dp2[i + 2], dp2[i - 1] + nums[i + 3]);
            }
        }
        return Math.max(dp1[n - 2], Math.max(dp2[n - 2], Math.max(dp1[n - 3], dp2[n - 3])));
    }
}
// @lc code=end

